### PR TITLE
Nested ldap queries

### PR DESCRIPTION
**Pull request by carneyweb**
This is a new algorithm for matching ldap values. It correctly deals with nested data. Performance appears to be equivalent to old flat implementation. Passes all tests.

Issuing pull request for review.

### CODE REVIEW

File: framework/test/bundles/libLQ/resources/manifest.json [hunk 1]
```diff
@@ -1,8 +1,23 @@
...
+    "bundle.symbolic_name" : "TestBundleLQ",
+    "bundle.testproperty" : "YES",
+    "bundle.nestedproperty" : {
+        "foo" : "bar"
+    },
+    "bundle.activator" : true,
+    "hosed" : "1",
+    "hosedd" : "2",
+    "hose" : "3",
+    "a" : {
```
Comment (jeffdiclemente) [inline framework/test/bundles/libLQ/resources/manifest.json:11]: would be useful to add test cases that test the order of search for a scenario like:

```
"a.b" : "foo",
"a" : { 
         "b" : "bar"
}
```

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]
This is good. There should be user facing documentation which describes the change in search behavior. This has the potential to break code searching on data like this:

```
"a.b" : "foo",
"a" : {
         "b" : "bar"
}
```
using a filter like "(a.b)==(bar)" and expecting a match. We can argue about how common or not that would be, however the fact remains that the search behavior has changed and needs to be documented for users.

File: framework/test/bundles/libLQ/resources/manifest.json (hunk 1 above)
Comment (carneyweb) [reply framework/test/bundles/libLQ/resources/manifest.json:11]: Done.

**Review by achristoforides — CHANGES_REQUESTED** [review]

File: framework/src/util/LDAPExpr.cpp [hunk 2]
```diff
@@ -31,21 +31,184 @@
...
+            // Loop through the input string and check each character for a delimiter. If a
+            // delimiter is found, add the string processed so far to the result container.
+            for (auto it = std::begin(input); it != std::end(input); std::advance(it, 1))
+            {
+                // If *it is a delimiter, the value marked by the delimiter is the string between
+                // "beg" and "it"... save it off into the result container.
+                if (delim[*it])
+                {
+                    if (in_token)
+                    {
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply framework/src/util/LDAPExpr.cpp:84]: That will actually change the behavior. In that case, if "denim[*it]" is set, and "in_token" is false, we'll end up in the else block, whereas with the original code we would not. The original code collapses sequential delimiters this way, and that would not happen if I made the above change. Going to leave it alone.

File: framework/src/util/LDAPExpr.cpp [hunk 3]
```diff
@@ -31,21 +31,184 @@
...
+            std::function<typename MapT::const_iterator(AnyMap const* p)> end_iter)
+        {
+            // short ciruit check. See if the full attrName is defined at the top level and return
+            // quickly if it is. We match this first to preserve existing behavior and only proceed
+            // to "walk down" the JSON tree if we don't find the value at the top level.
+            auto lookup = get_value_from_map(pPtr, attrName);
+            if (lookup != end_iter(pPtr))
+            {
+                return lookup;
+            }
```
[the remark this answers is not in the document]
Comment (carneyweb) [reply framework/src/util/LDAPExpr.cpp:138]: Done.

File: framework/src/util/LDAPExpr.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (achristoforides) [reply framework/src/util/LDAPExpr.cpp:84]: Ah good catch.

**Review by achristoforides — APPROVED** [review]
LGTM

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

File: framework/include/cppmicroservices/LDAPFilter.h [hunk 4]
```diff
@@ -59,6 +59,19 @@ namespace cppmicroservices
...
+     * at the top level for a matching entry, and if one isn't found, we decompose the nested key
+     * and "walk down" the JSON structure looking for a match. For example:
+     *
+     *   - "("bundle.symbolic_name=my_bundle)". This will look for a match in two places. First it
+     *     will look in the JSON manifest for:
+     *     { "bundle.symbolic_name" : "my_bundle" }
+     *
+     *     If not found there, this will be checked:
+     *     { "bundle" : { "symbolic_name" : "my_bundle" } }
+     *   - top level flat keys are preferred in order to preserve the behavior of existing filters.
```
[the remark this answers is not in the document]
Comment (jeffdiclemente) [reply framework/include/cppmicroservices/LDAPFilter.h:74]: FYI, if you make a comment change, you can skip running CI by adding ```[ci skip]``` to the commit message

**Review by jeffdiclemente — CHANGES_REQUESTED** [review]

**Review by jeffdiclemente — APPROVED** [review]

**Pull request merged**